### ISSUE TITLE
Rare crash when reformatting file:

### ISSUE THREAD

**Issue by jordwalke** [facebook/reason#860]
When the file contents are exactly as follows, parsing succeeds but printing fails. This should be a good first issue to help out with. It is likely a straightforward fix.

```
/*
 * vim: set ft=rust:
 * vim: set ft=reason:
 */

/**
 * Copyright (c) 2014-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
let floatMult = ( *. );

let floatDiv = (/.);

let floatSub = (-.);

open LayoutTestUtils;


/**
 * Since Core_bench is such a huge dependency and it doesn't compile with byte,
 * we include a fake shim of it that we enable by default. To use the far
 * superior Core_bench.
 *
 * - Comment out `include FakeCore;` below.
 * - Uncomment `open Core_bench.Std;`
 * - Delete the two targets in package.json (byteTarget, jsTarget)
 * - Run `npm run build`, then `npm run bench`
 *
 */
let module FakeCore = {
  let module Bench = {
    let module Test = {
      let create name::s itm => (s, itm);
    };
    let make_command listofCreatedBenchmarks => listofCreatedBenchmarks;
  };
  let module Core = {
    let runCommand (name, func) => {
      let numIterations = 100;
      let timesInMs = Array.make_float numIterations;
      Gc.full_major ();
      let startSeconds = Sys.time ();
      for i in 0 to (numIterations - 1) {
        let itemStartSeconds = Sys.time ();
        func ();
        let itemEndSeconds = Sys.time ();
        timesInMs.(i) = floatSub itemEndSeconds itemStartSeconds
      };
      let endSeconds = Sys.time ();
      print_string (
        "Average ms for " ^
        name ^
        " " ^
        string_of_float (
          floatMult 1000.0 (floatDiv (floatSub endSeconds startSeconds) (float_of_int numIterations))
        )
      );
      print_newline ();
      Array.sort compare timesInMs;
      print_string (
        "Median ms for " ^
        name ^
        " " ^
        string_of_float (floatMult 1000.0 timesInMs.(numIterations / 2)) ^ " (Not Valid For JS benchmarks)"
      );
      print_newline ()
    };
    let module Std = {
      let module Command = {
        let run listOftests => List.iter (fun command => runCommand command) listOftests;
      };
    };
  };
};

include FakeCore;

let something = ListLabels.exists;

/* static CSSSize _measure(void *context,
                           float width,
                           CSSMeasureMode widthMode,
                           float height,
                           CSSMeasureMode heightMode) {
     int *measureCount = (int *)context;
     *measureCount = *measureCount + 1;
     return CSSSize {
         .width = widthMode == CSSMeasureModeUndefined ? 10 : width,
         .height = heightMode == CSSMeasureModeUndefined ? 10 : width,
     };
   } */
it
  "should measure correctly"
  (
    fun () => {
      /* let module LayoutSystem =
        LayoutSystem.Create
          {}
          {};
      let measureCount = {contents: 0};
      let measure context width widthMode height heightMode => {
        measureCount.contents = measureCount.contents + 1;
        {
          LayoutTypes.width: widthMode === LayoutTypes.CssMeasureModeUndefined ? 10 : width,
          height: heightMode === LayoutTypes.CssMeasureModeUndefined ? 10 : height
        }
      };
      let root_style = {...LayoutSupport.defaultStyle, width: 10000, height: 10000};
      let root_child0_style = {
        ...LayoutSupport.defaultStyle,
        positionType: CssPositionAbsolute,
        flexGrow: 1,
        left: 2000,
        endd: 1000,
        right: 2000
      };
      let root_child0 =
        LayoutSupport.createNode withChildren::[||] andStyle::root_child0_style andMeasure::measure ();
      let root =
        LayoutSupport.createNode withChildren::[|root_child0|] andStyle::root_style andMeasure::measure ();
      Layout.layoutNode root cssUndefined cssUndefined CssDirectionLtr;
      if (
        root.layout.top != 0 ||
        root.layout.left != 0 ||
        root.layout.width != 10000 ||
        root.layout.height != 10000 ||
        root_child0.layout.top != 0 ||
        root_child0.layout.left != 2000 || root_child0.layout.width != 7000 || root_child0.layout.height != 0
      ) {
        assertLayouts
          170
          ({...root.layout, top: 0, left: 0, width: 10000, height: 10000}, root.layout)
          [({...root_child0.layout, top: 0, left: 2000, width: 7000, height: 0}, root_child0.layout)]
      };
      Layout.layoutNode root cssUndefined cssUndefined CssDirectionRtl;
      if (
        root.layout.top != 0 ||
        root.layout.left != 0 ||
        root.layout.width != 10000 ||
        root.layout.height != 10000 ||
        root_child0.layout.top != 0 ||
        root_child0.layout.left != 1000 || root_child0.layout.width != 7000 || root_child0.layout.height != 0
      ) {
        assertLayouts
          171
          ({...root.layout, top: 0, left: 0, width: 10000, height: 10000}, root.layout)
          [({...root_child0.layout, top: 0, left: 1000, width: 7000, height: 0}, root_child0.layout)]
      } */
    }
  );

LayoutTestUtils.displayOutcomes ();
```

**Comment by jordwalke** [comment]
The crash:

```
Fatal error: exception Failure("hd")
Raised at file "pervasives.ml", line 30, characters 22-33
Called from file "list.ml", line 23, characters 10-23
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1907, characters 21-58
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1892, characters 19-53
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1892, characters 19-53
Called from file "src/reason_pprint_ast.ml", line 1907, characters 21-58
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 1892, characters 19-53
Called from file "src/reason_pprint_ast.ml", line 1869, characters 25-60
Called from file "src/reason_pprint_ast.ml", line 2011, characters 18-56
Called from file "list.ml", line 84, characters 24-34
Called from file "src/reason_pprint_ast.ml", line 6018, characters 26-128
Called from file "src/refmt_impl.ml", line 235, characters 6-32
```

**Comment by jmmk** [comment]
Can reproduce `Fatal error: exception Failure("hd")` with:

```
fun () => { 
/**/
}
``` 

And a slight variation `Fatal error: exception Failure("tl")` with:

```
fun () => {/**/}
``` 
---

If we compare the AST of a working example with four spaces between braces (to match the amount of space taken up by the comment) `fun () => {    }` and the AST of one of our broken examples `fun () => {/**/}`, they are identical. 

Is there any information that is either not encoded in the AST, or not printed by the AST printer?

**Comment by jordwalke** [comment]
The printer deals with more than just the AST - it also interleaves comments. This would be a great first task for anyone interested. Feel free to try to fix it.

**Comment by jmmk** [comment]
@jordwalke so comments are parsed, but are not part of the AST? It looks like comments are maintained only when using `-print re`?

**Comment by jmmk** [comment]
I think I found the answer to my questions in `reason_toolchain.ml`

Lines 356-361 indicate that indeed we can't print comments in `.ml` files:

```ocaml
  (* Unfortunately we drop the comments because there doesn't exist an ML
   * printer that formats comments *and* line wrapping! (yet) *)
  let format_interface_with_comments (signature, _) formatter =
    Pprintast.signature formatter signature
  let format_implementation_with_comments (implementation, _) formatter =
    Pprintast.structure formatter implementation
```

And I see in the types that the parse tree (I'm assuming this is what `-print ast` gives us) is treated as separate from the comments (e.g. `(Parsetree.structure * Reason_pprint_ast.commentWithCategory)`)

**Comment by jmmk** [comment]
Found out where it's happening:

in `reason_pprint_ast.ml`, `looselyAttachComment`, in the match `| Sequence (listConfig, subLayouts)` (currently [line 1839](https://github.com/facebook/reason/blob/master/src/reason_pprint_ast.ml#L1839)).

`subLayouts` is an empty list, so the first statement returns a tuple `(beforeComment, afterComment)` where both of those values are also empty lists. We match on `List.rev beforeComment` which hits the first match `[]` and attempts to call `prependSingleLineComment comment (List.hd afterComment) :: (List.tl afterComment)`. This is where the failure occurs, because both `List.tl afterComment` and `List.hd afterComment` will fail.

**Comment by jmmk** [comment]
`print_layout` gives:

```
 [0 (1:0)-23 (1:23)]
   Sequence of 1, sep: ; stick_to_left: true break: Always_rec
     [0 (1:0)-23 (1:23)]
       [0 (1:0)-23 (1:23)]
         Label:
           left
           Sequence of 1, sep:  stick_to_left: true break: if need
             [4 (1:4)-6 (1:6)]
               [4 (1:4)-6 (1:6)]
                 [4 (1:4)-6 (1:6)]
                   Easy: ()
           right
           [10 (1:10)-23 (1:23)]
             [10 (1:10)-23 (1:23)]
               Sequence of 0, sep: ; stick_to_left: true break: if need
```

I just realized that the last Sequence with 0 nested `layoutNode`s is the function body. The failure seems to be specific to an empty function (possibly other empty blocks - are any other empty blocks valid language constructs?). Adding any code inside the function body (besides a comment) will not cause an error.

**Comment by jmmk** [comment]
Adding a clause for `| Sequence (listConfig, subLayouts) when subLayouts == [] -> formatComment txt` gets me close, but we need to enclose the comment in the braces. 

It looks like there are rules for braces depending on how many forms are inside, so we might need to override for this case.

**Comment by jmmk** [comment]
I think `| Sequence (listConfig, subLayouts) when subLayouts == [] -> Sequence (listConfig, [formatComment txt])` does the right thing, though I'm not 100% sure I'm using `Sequence` as intended. Is it meant for "blocks" meaning "section of code enclosed by curly braces"?

**Comment by jmmk** [comment]
@jordwalke opened PR #875

**Comment by jmmk** [comment]
@jordwalke @yunxing I think this can be closed since #875 was merged

**Comment by jordwalke** [comment]
Thanks for helping us organize the tasks!
